### ISSUE TITLE
[Code issues?] Possible code issues found by CppCheck

### ISSUE THREAD

**Issue by driekus77** [libgit2/libgit2#3350]
CppCheck returns some possible errors and warnings running on libgit2(Win x64 Release):

![libgit2_cppcheck_report](https://cloud.githubusercontent.com/assets/1574281/9047643/89110a5c-3a33-11e5-8b0d-1e25b800ca09.png)

Run it on today's master ( 3 aug 2015 at 22:00 GMT+1:00 (Amsterdam)) 
$ git describe
v0.16.0-6666-g69adb78

and 0.23.1 Release.
Both have approx. same error list.

Steps to reproduce:
   - Install CppCheck command-line tool: See website http://cppcheck.sourceforge.net
   - set PATH="c:\program files (x86)\cppcheck";%PATH%
   - cd libgit2/src
   - cppcheck --xml --inline-suppr -v --template="{file}|{line}|{severity}|{id}|{message}" --platform=win64 --library=windows --library=std --std=c89 --language=c *.c 2>CppCheckReport.xml

Results in CppCheckReport.xml:
```xml
<results>
    <error file="attr.c" line="362" id="memleak" severity="error" msg="Memory leak: macro"/>
    <error file="attr_file.c" line="454" id="uninitvar" severity="error" msg="Uninitialized variable: key"/>
    <error file="buffer.c" line="377" id="va_list_usedBeforeStarted" severity="error" msg="va_list &apos;args&apos; used before va_start() was called."/>
    <error file="buffer.c" line="383" id="va_list_usedBeforeStarted" severity="error" msg="va_list &apos;args&apos; used before va_start() was called."/>
    <error file="buffer.c" line="385" id="va_list_usedBeforeStarted" severity="error" msg="va_list &apos;args&apos; used before va_start() was called."/>
    <error file="buffer.c" line="542" id="va_end_missing" severity="error" msg="va_list &apos;ap&apos; was opened but not closed by va_end()."/>
    <error file="curl_stream.c" line="82" id="memleak" severity="error" msg="Memory leak: str"/>
    <error file="curl_stream.c" line="210" id="memleak" severity="error" msg="Memory leak: st"/>
    <error file="diff_print.c" line="378" id="memleak" severity="error" msg="Memory leak: ofile"/>
    <error file="diff_print.c" line="377" id="memleak" severity="error" msg="Memory leak: nfile"/>
    <error file="index.c" line="2001" id="memleak" severity="error" msg="Memory leak: conflict_name"/>
    <error file="iterator.c" line="599" id="memleak" severity="error" msg="Memory leak: root"/>
    <error file="pack.c" line="1072" id="uninitvar" severity="error" msg="Uninitialized variable: sha1"/>
    <error file="refs.c" line="215" id="nullPointer" severity="error" msg="Possible null pointer dereference: ref"/>
    <error file="refspec.c" line="327" id="memleak" severity="error" msg="Memory leak: cur"/>
    <error file="repository.c" line="2306" id="memleak" severity="error" msg="Memory leak: tmp_name"/>
    <error id="toomanyconfigs" severity="style" msg="The checking of the file will be interrupted because there are too many #ifdef configurations. Checking of all #ifdef configurations can be forced by --force command line option or from GUI preferences. However that may increase the checking time. For more details, use --enable=information."/>
</results>
```
Looking at some errors they look false positives but maybe someone with more experience with libgit2 can find a real issue?

Kind regards, 
Henry

**Comment by ethomson** [comment]
Hard to tell for sure.  It appears that most of the memory leak warnings are legitimate (from a code standpoint), but our policy is to bail quickly if an allocation fails and not to try to clean up a few bytes that were allocated earlier.  Though we think this makes a lot of sense practically speaking, it certainly angers static code analysis tools.

So this is creating a lot of noise.  There may be some valid errors in there (though there also appear to false positives).  If you are interested in using this tool, then I would suggest turning off memory leak detection and simply using the one built-in to the C runtime, or using a tool like Dr Memory, which actually shows memory leaks in regular usage instead of the ones detecting through code analysis.

**Comment by denji** [comment]
https://github.com/libgit2/libgit2/commit/252f86e1bef45c0376a119602ff4943b3c476084

warning,error,performance
```
[deps/regex/regcomp.c:3059]: (error) Memory leak: sbcset
[deps/regex/regcomp.c:3607]: (error) Memory leak: sbcset
[deps/regex/regcomp.c:532]: (error) Memory leak: fastmap
[deps/regex/regcomp.c:3059]: (error) Memory leak: mbcset
[deps/regex/regcomp.c:3607]: (error) Memory leak: mbcset
[deps/regex/regcomp.c:2775]: (error) Uninitialized variable: table_size
[deps/regex/regcomp.c:2776]: (error) Uninitialized variable: symb_table
[deps/regex/regcomp.c:2778]: (error) Uninitialized variable: table_size
[deps/zlib/trees.c:826]: (warning) %ld in format string (no. 1) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:826]: (warning) %ld in format string (no. 2) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:853]: (warning) %ld in format string (no. 1) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:856]: (warning) %ld in format string (no. 1) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:859]: (warning) %ld in format string (no. 1) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:926]: (warning) %ld in format string (no. 1) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:926]: (warning) %ld in format string (no. 2) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:930]: (warning) %ld in format string (no. 1) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:930]: (warning) %ld in format string (no. 2) requires 'long' but the argument type is 'unsigned long'.
[deps/zlib/trees.c:1090]: (warning) %d in format string (no. 1) requires 'int' but the argument type is 'unsigned int'.
[deps/zlib/trees.c:364]: (warning) %u in format string (no. 1) requires 'unsigned int' but the argument type is 'int'.
[deps/zlib/trees.c:370]: (warning) %u in format string (no. 1) requires 'unsigned int' but the argument type is 'int'.
[deps/zlib/zutil.c:278]: (warning) Assignment of function parameter has no effect outside the function. Did you forget dereferencing it?
[deps/zlib/zutil.c:284]: (warning) Assignment of function parameter has no effect outside the function. Did you forget dereferencing it?
[deps/zlib/zutil.c:259]: (warning) Assignment of function parameter has no effect outside the function. Did you forget dereferencing it?
[src/attr.c:362]: (error) Memory leak: macro
[src/attr_file.c:454]: (error) Uninitialized variable: key
[src/buffer.c:377]: (error) va_list 'args' used before va_start() was called.
[src/buffer.c:383]: (error) va_list 'args' used before va_start() was called.
[src/buffer.c:385]: (error) va_list 'args' used before va_start() was called.
[src/buffer.c:542]: (error) va_list 'ap' was opened but not closed by va_end().
[src/checkout.c:736] -> [src/checkout.c:733]: (warning) Possible null pointer dereference: a - otherwise it is redundant to check it against null.
[src/cache.c:53]: (warning) %Id in format string (no. 3) requires 'ptrdiff_t' but the argument type is 'ssize_t {aka long}'.
[src/curl_stream.c:82]: (error) Memory leak: str
[src/curl_stream.c:210]: (error) Memory leak: st
[src/diff.c:1012] -> [src/diff.c:1015]: (performance) Variable 'recurse_into_dir' is reassigned a value before the old one has been used.
[src/diff_print.c:378]: (error) Memory leak: ofile
[src/diff_print.c:377]: (error) Memory leak: nfile
[src/index.c:2001]: (error) Memory leak: conflict_name
[src/iterator.c:1806] -> [src/iterator.c:1810]: (warning) Possible null pointer dereference: entry - otherwise it is redundant to check it against null.
[src/iterator.c:599]: (error) Memory leak: root
[src/pack.c:1072]: (error) Uninitialized variable: sha1
[src/pack.c:1348]: (warning) %d in format string (no. 1) requires 'int' but the argument type is 'unsigned int'.
[src/refs.c:215]: (error) Possible null pointer dereference: ref
[src/refs.c:1099] -> [src/refs.c:1088]: (warning) Possible null pointer dereference: ref - otherwise it is redundant to check it against null.
[src/refspec.c:327]: (error) Memory leak: cur
[src/repository.c:2306]: (error) Memory leak: tmp_name
[src/stransport_stream.c:181]: (portability) 'data' is of type 'void *'. When using void pointers in calculations, the behaviour is undefined.
[src/win32/path_w32.c:195]: (error) Buffer is accessed out of bounds.
[src/win32/path_w32.c:211]: (error) Buffer is accessed out of bounds.
[src/win32/posix_w32.c:388]: (error) Uninitialized variable: target_w
[src/win32/w32_crtdbg_stacktrace.c:95]: (warning) %lx in format string (no. 1) requires 'unsigned long' but the argument type is 'unsigned int'.
[src/win32/w32_crtdbg_stacktrace.c:133]: (warning) %lx in format string (no. 1) requires 'unsigned long *' but the argument type is 'unsigned int *'.
[src/win32/w32_crtdbg_stacktrace.c:190]: (warning) %d in format string (no. 1) requires 'int' but the argument type is 'unsigned int'.
[src/win32/w32_crtdbg_stacktrace.c:190]: (warning) %d in format string (no. 2) requires 'int' but the argument type is 'unsigned int'.
[src/win32/w32_crtdbg_stacktrace.c:190]: (warning) %d in format string (no. 3) requires 'int' but the argument type is 'unsigned int'.
[src/win32/w32_crtdbg_stacktrace.c:194]: (warning) %d in format string (no. 1) requires 'int' but the argument type is 'unsigned int'.
[src/win32/w32_crtdbg_stacktrace.c:194]: (warning) %d in format string (no. 2) requires 'int' but the argument type is 'unsigned int'.
[src/win32/w32_crtdbg_stacktrace.c:202]: (warning) %d in format string (no. 2) requires 'int' but the argument type is 'unsigned int'.
[src/win32/w32_crtdbg_stacktrace.c:202]: (warning) %d in format string (no. 3) requires 'int' but the argument type is 'unsigned int'.
[tests/clar.c:545] -> [tests/clar.c:538]: (warning) Possible null pointer dereference: s1 - otherwise it is redundant to check it against null.
[tests/clar.c:545] -> [tests/clar.c:538]: (warning) Possible null pointer dereference: s2 - otherwise it is redundant to check it against null.
[tests/clar.c:563] -> [tests/clar.c:556]: (warning) Possible null pointer dereference: s1 - otherwise it is redundant to check it against null.
[tests/clar.c:563] -> [tests/clar.c:556]: (warning) Possible null pointer dereference: s2 - otherwise it is redundant to check it against null.
[tests/core/string.c:51]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:52]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:53]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:54]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:55]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:72]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:73]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:74]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:75]: (warning) Unnecessary comparison of static strings.
[tests/core/string.c:76]: (warning) Unnecessary comparison of static strings.
[tests/diff/notify.c:184] -> [tests/diff/notify.c:188]: (performance) Variable 'pathspec' is reassigned a value before the old one has been used.
[tests/online/clone.c:396] -> [tests/online/clone.c:400]: (performance) Variable 'with_user' is reassigned a value before the old one has been used.
[tests/repo/init.c:577] -> [tests/repo/init.c:593]: (performance) Variable 'filemode' is reassigned a value before the old one has been used.
```

**Comment by linquize** [comment]
But usually we avoid modifying 3rd party source such as regex library

**Comment by carlosmn** [comment]
I've looked into a bunch of these warnings, but they all seem to stem from the analyzer not understanding what we're doing. The missing opens for `va_list` just mean that we opened in a different function,the reassignment warnings do not take into consideration that we use the *address* of the variable, so it doesn't notice that the functions are in fact using it, and so forth.

Most of the warnings are for external code as well, or for leaks which we don't care about as they'd be in the OOM path; so I'm going to close this as it doesn't seem to show actions we could in fact take. If valgrind/safe stack/address sanitizer does show warnings, that's something we can work with.
